File: 2018/11/python/solution.py
#!/usr/bin/env pypy3
from collections import namedtuple
from itertools import chain
from operator import attrgetter

GRID_SIZE = 300

INPUT = 7672


Window = namedtuple("Window", "x, y, window_size, total_power")


def windowed(grid, window_size):
    y = len(grid)
    x = len(grid[0])
    for i in range(1, x - window_size):
        for j in range(1, y - window_size):
            a = grid[j][i]
            b = grid[j][i + window_size]
            c = grid[j + window_size][i + window_size]
            d = grid[j + window_size][i]
            power_level = c + a - d - b
            yield Window(i + 1, j + 1, window_size, power_level)


def power_level(x, y, serial_number):
    rack_id = x + 10
    return ((((rack_id * y + serial_number) * rack_id) // 100) % 10) - 5


def calculate_power_level(serial_number):
    return [
        [power_level(x, y, serial_number) for x in range(GRID_SIZE + 1)]
        for y in range(GRID_SIZE + 1)
    ]


def precompute_sums(grid):
    for line in grid:
        for x in range(1, len(line)):
            line[x] += line[x - 1]
    for y in range(1, len(grid)):
        for x in range(len(grid[y])):
            grid[y][x] += grid[y - 1][x]


def solve(power_levels, window_sizes):
    return max(
        chain.from_iterable(
            windowed(power_levels, window_size)
            for window_size in window_sizes
        ),
        key=attrgetter("total_power")
    )


def main():
    power_levels = calculate_power_level(INPUT)
    precompute_sums(power_levels)
    x, y, _, _ = solve(power_levels, [3])
    print(f"{x},{y}")
    x, y, window_size, _ = solve(power_levels, range(1, GRID_SIZE + 1))
    print(f"{x},{y},{window_size}")


if __name__ == "__main__":
    main()

File: 2018/11/cpp/solution.cxx
#include <cstdint>

#include <array>
#include <algorithm>
#include <iostream>
#include <iterator>
#include <numeric>
#include <optional>
#include <vector>


constexpr auto GRID_SIZE = size_t{300};
constexpr auto INPUT = int64_t{7672};

using Grid = std::vector<std::vector<int64_t>>;

struct Window {
    size_t x;
    size_t y;
    size_t window_size;
    int64_t total_power;
};


auto power_level(size_t const x, size_t const y, int64_t const serial_number) -> int64_t {
    auto const rack_id = static_cast<int64_t>(x) + 10;
    return ((((rack_id * static_cast<int64_t>(y) + serial_number) * rack_id) / 100) % 10) - 5;
}

auto calculate_grid(int64_t serial_number) -> Grid {
    auto grid = Grid{GRID_SIZE + 1, std::vector<int64_t>(GRID_SIZE + 1, 0)};
    for (auto y = size_t{0}; y < grid.size(); ++y) {
        for (auto x = size_t{0}; x < grid[y].size(); ++x) {
            grid[y][x] = power_level(x, y, serial_number);
        }
    }
    return grid;
}

auto precompute_sums(Grid& grid) -> void {
    for (auto& line: grid) {
        for (auto x = size_t{1}; x < line.size(); ++x) {
            line[x] += line[x - 1];
        }
    }
    for (auto y = size_t{1}; y < grid.size(); ++y) {
        for (auto x = size_t{0}; x < grid[y].size(); ++x) {
            grid[y][x] += grid[y - 1][x];
        }
    }
}

template<typename OutputIt>
auto windowed(Grid const& grid, size_t window_size, OutputIt out) -> void {
    auto const y = grid.size();
    auto const x = grid.at(0).size();
    for (auto i = size_t{1}; i < x - window_size; ++i) {
        for (auto j = size_t{1}; j < y - window_size; ++j) {
            auto const a = grid[j][i];
            auto const b = grid[j][i + window_size];
            auto const c = grid[j + window_size][i + window_size];
            auto const d = grid[j + window_size][i];
            auto const power_level = c + a - d - b;
            *out++ = Window{i + 1, j + 1, window_size, int64_t{power_level}};
        }
    }
}

template<typename T, auto compare = std::less<T>{}>
struct Maximum {
    using value_type = T;

    std::optional<T> maximum;

    auto push_back(T&& t) -> void {
        if (not maximum or compare(*maximum, t)) {
            maximum.emplace(std::forward<T>(t));
        }
    }
};

template<typename It>
auto solve(Grid const& grid, It const first, It const last) -> Window {
    auto solution = Maximum<
        Window,
        [](auto const& lhs, auto const& rhs) { return lhs.total_power <= rhs.total_power; }
    >{};
    auto inserter = std::back_inserter(solution);
    std::for_each(first, last, [&](auto const& window_size) {
        windowed(grid, window_size, inserter);
    });
    return *solution.maximum;
}

auto main() -> int {
    auto const power_levels = []() {
        auto ps = calculate_grid(INPUT);
        precompute_sums(ps);
        return ps;
    }();
    auto const grid_sizes = []() {
        auto grid_sizes = std::array<size_t, GRID_SIZE + 1>{};
        std::iota(begin(grid_sizes), end(grid_sizes), 0);
        return grid_sizes;
    }();
    auto const part1 = solve(power_levels, begin(grid_sizes) + 3, begin(grid_sizes) + 4);
    std::cout << part1.x << ',' << part1.y << '\n';
    auto const part2 = solve(power_levels, begin(grid_sizes), end(grid_sizes));
    std::cout << part2.x << ',' << part2.y << ',' << part2.window_size << '\n';
}
